R/DmmParam.R

```diff
@@ -0,0 +1,196 @@
+#' Dirichlet Multinomial Mixtures clustering
+#' 
+#' Use the DMM algorithm from the \code{DirichletMultinomial} package on a 
+#' distance matrix within \code{\link{clusterRows}}.
+#' 
+#' @param k An integer vector indicating the number of clusters to choose from (or the possible number of clusters if the vector has multiple values)
+#' @param type A string specifying the fit function to use to find the optimal number of clusters. 
+#'   Use in combination with \code{k} being a list of integers. 
+#'   Must be equal to \code{"laplace"}, \code{"AIC"} or \code{"BIC"}.
+#' @param transposed Logical scalar, is x transposed with samples in rows?
+#' @param seed An integer specifying the seed to use when doing the DMM 
+#'   algorithm. Will be random if not set.
+#' 
+#' @inheritParams clusterRows
+#' 
+#' @author Basil Courbayre
+#'
+#' @references
+#' Holmes I, Harris K and Quince C (2012).
+#' Dirichlet multinomial mixtures: generative models for microbial metagenomics.
+#' \emph{PLoS ONE, vol. 7(2)}, 1-15
+#' 
+#' @details
+#' The DMM algorithm (see Holmes et al. 2012) is commonly used in microbial 
+#' ecology along with metagenomic and 16S rRNA count data.
+#' 
+#' To modify an existing DmmParam object \code{x},
+#' users can simply call \code{x[[i]]} or \code{x[[i]] <- value} where \code{i} 
+#' is any argument used in the constructor.
+#'
+#' The default for \code{k} is \code{1:3}.
+#' 
+#' The default for \code{type} is \code{"laplace"}.
+#' 
+#' @return 
+#' The \code{DmmParam} constructor will return a \linkS4class{DmmParam} object 
+#' with the specified parameters.
+#'
+#' The \code{clusterRows} method will return a factor of length equal to 
+#' \code{nrow(x)} containing the cluster assignments.
+#' If \code{full=TRUE}, a list is returned with \code{clusters} 
+#' (the factor, as above) and \code{objects} 
+#' (a list containing \code{dmm}, the output of 
+#' \code{\link[DirichletMultinomial]{dmn}} for each value contained in 
+#' \code{k}; \code{k}, the number of clusters; \code{prob}, the array of 
+#' probabilities; and \code{seed} the seed for the dmm).
+#'
+#' @examples
+#' clusterRows(iris[,1:4], DmmParam())
+#' clusterRows(iris[,1:4], DmmParam(k=2))
```

Comment:
Actually, the iris examples are surprisingly expensive:

```r
system.time(clusterRows(iris[,1:4], DmmParam(k=2)))
##    user  system elapsed
##  17.451   0.044  17.500

system.time(clusterRows(iris[,1:4], DmmParam()))
##    user  system elapsed
##  43.611   0.118  43.781


system.time(clusterRows(round(counts), DmmParam(k=1:3, type="laplace")))
##    user  system elapsed
##   1.790   0.005   1.794
```

I don't know why that is, but we need to get this below a few seconds.